File: scala/src/test/scala/wordle/internal/PuzzleTest.scala
package wordle.internal

import org.mockito.ArgumentMatchers.any
import wordle.internal.Puzzle
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatestplus.mockito.MockitoSugar
import org.mockito.Mockito.*
import org.scalatest.BeforeAndAfterEach
import org.scalatest.matchers.should.Matchers
import wordle.internal.domain.{Correct, Display, GuessFeedback, InProgress, IncorrectLetter, IncorrectLocation, LetterFeedback, Lost, Player, Won}

class PuzzleTest extends AnyFlatSpec with MockitoSugar with Matchers with BeforeAndAfterEach {

  private val mockPlayer: Player = mock[Player]

  override protected def beforeEach(): Unit = {
    reset(mockPlayer)
  }

  behavior of "createPuzzle"

  it should "create a puzzle with the initial state" in {
    // Given
    val wordList = List("abcde", "bcdef", "invalid-word")

    // When
    val actual = Puzzle.newPuzzle(wordList, mockPlayer)

    // Then
    actual match {
      case Puzzle(wordList, player, word, puzzleStatus, guessFeedback) =>
        wordList shouldBe List("ABCDE", "BCDEF")
        player shouldBe mockPlayer
        List("ABCDE", "BCDEF") should contain (word)
        puzzleStatus shouldBe InProgress
        guessFeedback shouldBe List()
    }
  }

  behavior of "nextState"

  it should "return feedback for letter locations" in {
    // Given
    val puzzle = createPuzzle(List("abcde", "axyzb"), "abcde")
    when(mockPlayer.guessWord(any(classOf[Seq[GuessFeedback]]))).thenReturn("axyzb")

    val expectedLetterFeedback = Seq(
      LetterFeedback('A', Correct),
      LetterFeedback('X', IncorrectLetter),
      LetterFeedback('Y', IncorrectLetter),
      LetterFeedback('Z', IncorrectLetter),
      LetterFeedback('B', IncorrectLocation))

    // When
    val actual = puzzle.nextState()

    actual.guessFeedback.length shouldBe 1
    actual.guessFeedback.head.letterFeedback shouldBe expectedLetterFeedback
  }

  it should "lose a game with six failed guesses" in {
    // Given
    val puzzle = createPuzzle(List("abcde", "axyzb"), "abcde")
    when(mockPlayer.guessWord(any(classOf[Seq[GuessFeedback]]))).thenReturn("axyzb")

    // When
    val actual = puzzle.nextState().nextState().nextState().nextState().nextState().nextState()

    // Then
    actual.puzzleStatus shouldBe Lost
  }

  it should "win a game with a correct guess" in {
    // Given
    val puzzle = createPuzzle(List("abcde", "axyzb"), "abcde")
    when(mockPlayer.guessWord(any(classOf[Seq[GuessFeedback]]))).thenReturn("abcde")

    // When
    val actual = puzzle.nextState()

    // Then
    actual.puzzleStatus shouldBe Won
  }

  it should "win a game with a correct guess in the final turn" in {
    // Given
    val puzzle = createPuzzle(List("abcde", "axyzb"), "abcde")
    when(mockPlayer.guessWord(any(classOf[Seq[GuessFeedback]])))
      .thenReturn("axyzb")
      .thenReturn("axyzb")
      .thenReturn("axyzb")
      .thenReturn("axyzb")
      .thenReturn("axyzb")
      .thenReturn("abcde")

    // When
    val actual = puzzle.nextState().nextState().nextState().nextState().nextState().nextState()

    // Then
    actual.puzzleStatus shouldBe Won
  }

  private def createPuzzle(wordList: List[String], word: String): Puzzle = {
    Puzzle.newPuzzle(wordList, mockPlayer).copy(word = word.toUpperCase)
  }
}

File: java/app/src/test/java/wordle/internal/PuzzleTest.java
package wordle.internal;

import com.google.common.collect.Lists;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import testcommon.MockTestBase;

import java.util.List;
import java.util.stream.IntStream;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.Mockito.when;
import static wordle.internal.Feedback.*;
import static wordle.internal.PuzzleStatus.LOST;
import static wordle.internal.PuzzleStatus.WON;

public class PuzzleTest extends MockTestBase {

    private final String correctWord = "TONER";
    private final String incorrectWord = "VOTER";
    private final List<String> wordList = Lists.newArrayList(correctWord, incorrectWord);

    @Mock private Player mockPlayer;

    @Test
    public void nextStateGivenIncorrectGuessThenProvidesFeedback() {
        // Given
        when(mockPlayer.guessWord(anyList())).thenReturn(incorrectWord);
        final Puzzle puzzle = new Puzzle(wordList, mockPlayer, correctWord);

        // When
        puzzle.nextState();

        // Then
        final List<GuessFeedback> actual = puzzle.getGuessFeedback();
        assertThat(actual.size()).isEqualTo(1);

        final List<LetterFeedback> expected = List.of(
                new LetterFeedback('V', INCORRECT_LETTER),
                new LetterFeedback('O', CORRECT),
                new LetterFeedback('T', INCORRECT_LOCATION),
                new LetterFeedback('E', CORRECT),
                new LetterFeedback('R', CORRECT)
        );

        assertThat(actual.get(0).getLetterFeedback())
                .usingRecursiveFieldByFieldElementComparator()
                .containsExactlyInAnyOrderElementsOf(expected);
    }

    @Test
    public void nextStateGivenTooManyIncorrectGuessesThenPuzzleIsLost() {
        // Given
        when(mockPlayer.guessWord(anyList())).thenReturn(incorrectWord);
        final Puzzle puzzle = new Puzzle(wordList, mockPlayer, correctWord);

        // When
        IntStream.rangeClosed(1, 6).forEach(x -> puzzle.nextState());

        // Then
        assertThat(puzzle.getStatus()).isEqualTo(LOST);
    }

    @Test
    public void nextStateGivenPuzzleWonThenNoChange() {
        // Given
        when(mockPlayer.guessWord(anyList())).thenReturn(correctWord);
        final Puzzle puzzle = new Puzzle(wordList, mockPlayer, correctWord);

        // When
        IntStream.rangeClosed(1, 2).forEach(x -> puzzle.nextState());

        // Then
        assertThat(puzzle.getStatus()).isEqualTo(WON);
    }

    @Test
    public void nextStateGivenCorrectGuessThenPuzzleIsWon() {
        // Given
        when(mockPlayer.guessWord(anyList())).thenReturn(correctWord);
        final Puzzle puzzle = new Puzzle(wordList, mockPlayer, correctWord);

        // When
        puzzle.nextState();

        // Then
        assertThat(puzzle.getStatus()).isEqualTo(WON);
    }

    @Test
    public void nextStateGivenPuzzleLostThenNoChange() {
        // Given
        when(mockPlayer.guessWord(anyList())).thenReturn(incorrectWord);
        final Puzzle puzzle = new Puzzle(wordList, mockPlayer, correctWord);

        // When
        IntStream.rangeClosed(1, 7).forEach(x -> puzzle.nextState());

        // Then
        assertThat(puzzle.getStatus()).isEqualTo(LOST);
    }
}
